pkg/nodeset/v1alpha1/types.go

```diff
@@ -0,0 +1,192 @@
+	// Template is the template or content of the file
+	// +optional
+	Template string `json:"template,omitempty"`
+
+	// For Reference type
+	// Reference points to an Object in registry that the controller will use when
+	// provisioning the node
+	// +optional
+	Reference *v1.ObjectReference `json:"reference,omitempty"`
+}
+
+// NodeClass describes the parameters for a class of nodes that can be
+// provisioned by the specific controller
+type NodeClass struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// NodeController specifes which node controller should provision and
+	// control this class of nodes
+	NodeController string `json:"nodeController,omitempty"`
+
+	// Config is free form data containing overridable parameters that the controller
+	// will use
+	// +optional
+	Config string `json:"config,omitempty"`
+
+	// Resources is a list of file-like resources that the controller can use
+	// +optional
+	Resources []NodeClassResource `json:"resources,omityemty"`
+}
+
+// NodeClassList is a collection of node classes.
+type NodeClassList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of NodeClasses
+	Items []NodeClass `json:"items"`
+}
+
+type NodeSetSpec struct {
+	// NodeSelector is a selector of node labels which a node must have to be in the NodeSet.
+	// More info: http://kubernetes.io/docs/user-guide/node-selection/README
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// NodeClass specifies which NodeClass the nodes created by this NodeSet should
+	// belong to
+	NodeClass string `json:"nodeClass,omitempty"`
+
+	// NodeSetController specifies a NodeSet controller to be in charge of this NodeSet
+	// If not specified, the default one that comes with the NodeController in NodeClass
+	// will be used.
+	// +optional
+	NodeSetController string `json:"nodeSetController,omityempty"`
+
+	// Replicas is the number of desired replicas.
+	Replicas int32
+
+	// The maximum number of Nodes that can be created/deleted at a time when
+	// reconciliation takes place. This can be an absolute number (ex. 3) or a percentage
+	// of total nodes at the time of reconciliation. The absolute number will be
+	// calculated from percentage by round up.
+	// By default, a value of 1 is used
+	// +optional
+	BatchSize intstr.IntOrString
+
+	// The maximum number of Nodes that can be created above the original number of
+	// Nodes during rolling-update. This can be an absolute number or a percentage of
+	// total nodes at the time of reconciliation.
+	// A default of 1 is used if not set.
+	// +optional
+	OverCapacity intstr.IntOrString
```

Comment:
Deployment uses MaxUnavailable and MaxSurge, similarly. The type should be a pointer. I missed that.